sdks/python/hatchet_sdk/worker/runner/runner.py

```diff
@@ -292,6 +413,164 @@ async def async_wrapped_action_func(
+            return
+
+        if not action.batch_id:
+            logger.error("Received START_BATCH without batch_id")
+            return
+
+        expected_size = 0
+        trigger_reason = None
+        trigger_time = None
+        batch_key = action.batch_key
+
+        if action.batch_start is not None:
+            expected_size = action.batch_start.expected_size
+            trigger_reason = action.batch_start.trigger_reason
+            trigger_time = action.batch_start.trigger_time
+
+        default_batch_size = (
+            controller.task.batch.batch_size  # type: ignore[union-attr]
+            if controller.task.batch is not None
+            else expected_size
+        )
+
+        controller.start_batch(
+            batch_id=action.batch_id,
+            expected_size=expected_size,
+            default_batch_size=default_batch_size,
+            trigger_reason=trigger_reason,
+            trigger_time=trigger_time,
+            batch_key=batch_key,
+            on_ready=lambda batch_id: asyncio.create_task(
+                self._maybe_flush_batch(controller, batch_id)
+            ),
+        )
+
+    async def _await_batch_item(
+        self,
+        ctx: Context,
+        task: Task[TWorkflowInput, R],
+        action: Action,
+    ) -> R:
+        controller = self._batch_controllers.get(action.action_id)
+        if controller is None:
+            raise RuntimeError(
+                f"Batch task '{action.action_id}' received batch metadata but no controller is registered"
+            )
+
+        if (
+            not action.batch_id
+            or action.batch_size is None
+            or action.batch_index is None
+        ):
+            raise RuntimeError(
+                f"Batch metadata missing for task '{action.action_id}', cannot process batch"
+            )
+
+        batch_id = action.batch_id
+        expected_size = action.batch_size
+        index = int(action.batch_index)
+
+        if expected_size <= 0:
+            raise RuntimeError(
+                f"Batch task '{action.action_id}' received invalid batch size: {expected_size}"
+            )
+
+        if index < 0:
+            raise RuntimeError(
+                f"Batch task '{action.action_id}' received invalid batch index: {index}"
+            )
+
+        loop = asyncio.get_running_loop()
+        future: asyncio.Future[Any] = loop.create_future()
+
+        controller.enqueue_item(
+            batch_id=batch_id,
+            expected_size=expected_size,
+            index=index,
+            ctx=ctx,
+            future=future,
+            batch_key=action.batch_key,
+            on_ready=lambda batch_id: asyncio.create_task(
+                self._maybe_flush_batch(controller, batch_id)
+            ),
+        )
+
+        try:
+            result = await future
+        except asyncio.CancelledError:
+            controller.remove_item(batch_id=batch_id, index=index)
+            raise
```

Comment:
this seems like the meat of it - how does this work? we create a bunch of futures, add callbacks for each item in the batch, and then try to await them resolving?